Firestore/core/src/firebase/firestore/util/executor_std.h

```diff
@@ -0,0 +1,283 @@
+                     [this, until] { return scheduled_.front().due != until; });
+      // There are 3 possibilities why `wait_until` has returned:
+      // - `wait_until` has timed out, in which case the current time is at
+      //   least `until`, so there must be an overdue entry;
+      // - a new entry has been added which comes before `until`. It must be
+      //   either overdue (in which case `HasDue` will break the cycle), or else
+      //   `until` must be reevaluated (on the next iteration of the loop);
+      // - `until` entry has been removed. This means `until` has to be
+      //   reevaluated, similar to #2.
+
+      if (HasDue()) {
+        return Extract(scheduled_.begin());
+      }
+    }
+  }
+
+  bool empty() const {
+    std::lock_guard<std::mutex> lock{mutex_};
+    return scheduled_.empty();
+  }
+
+  size_t size() const {
+    std::lock_guard<std::mutex> lock{mutex_};
+    return scheduled_.size();
+  }
+
+  // Removes the first entry satisfying predicate from the queue and returns it.
+  // If no such entry exists, returns an empty `optional`. Predicate is applied
+  // to entries in order according to their scheduled time.
+  //
+  // Note that this function doesn't take into account whether the removed entry
+  // is past its due time.
+  template <typename Pred>
+  absl::optional<T> RemoveIf(const Pred pred) {
+    std::lock_guard<std::mutex> lock{mutex_};
+
+    const auto found =
+        std::find_if(scheduled_.begin(), scheduled_.end(),
+                     [&pred](const Entry& s) { return pred(s.value); });
+    if (found != scheduled_.end()) {
+      return Extract(found);
+    }
+    return {};
+  }
+
+  // Checks whether the queue contains an entry satisfying the given predicate.
+  template <typename Pred>
+  bool Contains(const Pred pred) const {
+    std::lock_guard<std::mutex> lock{mutex_};
+
+    // While this logically duplicates part of the implementation of `RemoveIf`,
+    // a helper function would need two overloads, const and non-const,
+    // eliminating any gains in brevity.
+    return std::find_if(scheduled_.begin(), scheduled_.end(),
+                        [&pred](const Entry& s) { return pred(s.value); }) !=
+           scheduled_.end();
+  }
+
+ private:
+  struct Entry {
+    bool operator<(const Entry& rhs) const {
+      return due < rhs.due;
+    }
+
+    T value;
+    TimePoint due;
+  };
+  // All removals are on the front, but most insertions are expected to be on
+  // the back.
+  using Container = std::deque<Entry>;
+  using Iterator = typename Container::iterator;
+
+  void InsertPreservingOrder(Entry&& new_entry) {
+    std::lock_guard<std::mutex> lock{mutex_};
+
+    const auto insertion_point =
+        std::upper_bound(scheduled_.begin(), scheduled_.end(), new_entry);
+    scheduled_.insert(insertion_point, std::move(new_entry));
+
+    cv_.notify_one();
+  }
+
+  // This function expects the mutex to be already locked.
+  bool HasDue() const {
```

Comment:
You might consider naming this function in a way that suggests the lock is held (e.g. `HasDueLocked`).